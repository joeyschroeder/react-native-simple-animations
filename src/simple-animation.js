import React, { Component } from 'react';
import { Animated } from 'react-native';
import { getSpringAnimation, getTimingAnimation } from './util/get-animation';

import { AIMS } from './constants/aims';
import { DEFAULT_PROPS } from './constants/default-props';
import { DIRECTIONS } from './constants/directions';
import { MOVEMENT_TYPES } from './constants/movement-types';
import { PROP_TYPES } from './constants/prop-types';
import { STATIC_TYPES } from './constants/static-types';

export class SimpleAnimation extends Component {
  static propTypes = PROP_TYPES;

  static defaultProps = DEFAULT_PROPS;

  constructor(props) {
    super(props);

    this.opacityAnimation = new Animated.Value(0);
    this.movementAnimation = new Animated.Value(0);
    this.staticAnimation = new Animated.Value(0);
  }

  componentDidMount() {
    this.animate();
  }

  componentDidUpdate() {
    const { animateOnUpdate } = this.props;
    if (animateOnUpdate) this.animate();
  }

  getDistanceStartValue() {
    const { direction, distance } = this.props;

    switch (direction) {
      case DIRECTIONS.DOWN:
        return -distance;
      case DIRECTIONS.LEFT:
        return distance;
      case DIRECTIONS.RIGHT:
        return -distance;
      case DIRECTIONS.UP:
        return distance;

      default:
        return 0;
    }
  }

  getOpacity() {
    const { fade } = this.props;

    if (!fade) return {};
    return { opacity: this.opacityAnimation };
  }

  getMovementTransform() {
    const { direction, movementType } = this.props;

    if (!direction || !movementType) return [];

    const outputRangeStartValue = this.getDistanceStartValue();

    const interpolatedAnimation = this.movementAnimation.interpolate({
      inputRange: [0, 1],
      outputRange: [outputRangeStartValue, 0],
    });

    const transform = [
      direction === DIRECTIONS.LEFT || direction === DIRECTIONS.RIGHT
        ? { translateX: interpolatedAnimation }
        : { translateY: interpolatedAnimation },
    ];

    return transform;
  }

  getStaticTransform() {
    const { staticType } = this.props;

    if (!staticType) return [];
    return [{ scale: this.staticAnimation }];
  }

  getTransform() {
    const movementTransform = this.getMovementTransform();
    const staticTransform = this.getStaticTransform();

    const transform = [].concat(movementTransform, staticTransform);

    return { transform };
  }

  animate() {
    const { aim, animate, fade, movementType, staticType } = this.props;

    if (!animate) return;

    const fromValue = aim === AIMS.IN ? 0 : 1;
    const toValue = aim === AIMS.IN ? 1 : 0;

    if (fade) {
      const opacityAnimation = getTimingAnimation;
      this.opacityAnimation.setValue(fromValue);
      opacityAnimation({ animation: this.opacityAnimation, toValue, ...this.props }).start();
    }

    if (movementType) {
      const movementAnimation = movementType === MOVEMENT_TYPES.SPRING ? getSpringAnimation : getTimingAnimation;
      this.movementAnimation.setValue(fromValue);
      movementAnimation({ animation: this.movementAnimation, toValue, ...this.props }).start();
    }

    if (staticType) {
      const staticAnimation = staticType === STATIC_TYPES.BOUNCE ? getSpringAnimation : getTimingAnimation;
      this.staticAnimation.setValue(fromValue);
      staticAnimation({ animation: this.staticAnimation, toValue, ...this.props }).start();
    }
  }

  render() {
    const { children, style } = this.props;

    if (!children) return null;

    const viewStyles = [style, this.getTransform(), this.getOpacity()];

    return <Animated.View style={viewStyles}>{children}</Animated.View>;
  }
}
