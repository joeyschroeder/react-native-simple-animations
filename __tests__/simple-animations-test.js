import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Text } from 'react-native';
import { SimpleAnimation } from '../src/simple-animation';

jest.useFakeTimers();
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

describe('<SimpleAnimation />', () => {
  it('should call this.animate() when this.componentDidMount() is called', () => {
    const component = shallow(<SimpleAnimation />);
    const animate = jest.fn();

    component.instance().animate = animate;
    component.instance().componentDidMount();

    expect(animate).toHaveBeenCalled();
  });

  it('should not call this.animate() when this.componentDidUpdate() is called and props.animatedOnUpdate !== true', () => {
    const component = shallow(<SimpleAnimation animateOnUpdate={false} />);
    const animate = jest.fn();

    component.instance().animate = animate;
    component.instance().componentDidUpdate();

    expect(animate).not.toHaveBeenCalled();
  });

  it('should call this.animate() when this.componentDidUpdate() is called and props.animatedOnUpdate === true', () => {
    const component = shallow(<SimpleAnimation animateOnUpdate />);
    const animate = jest.fn();

    component.instance().animate = animate;
    component.instance().componentDidUpdate();

    expect(animate).toHaveBeenCalled();
  });

  it('should return 0 when this.getDistanceStartValue() is called and props.direction and props.distance to not exist', () => {
    const component = shallow(<SimpleAnimation />);
    const result = component.instance().getDistanceStartValue();

    expect(result).toEqual(0);
  });

  it('should return undefined when this.animate() is called and props.animate !== true', () => {
    const component = shallow(<SimpleAnimation animate={false} />);
    expect(component.instance().animate()).toEqual(undefined);
  });

  it('should return undefined when this.animate() is called and props.animate !== true', () => {
    const component = shallow(<SimpleAnimation animate={false} />);
    expect(component.instance().animate()).toEqual(undefined);
  });

  it('should return empty object when this.getOpacity() is called and props.fade !== true', () => {
    const component = shallow(<SimpleAnimation fade={false} />);
    expect(component.instance().getOpacity()).toEqual({});
  });

  it('should return object with opacity value when this.getOpacity() is called and props.fade === true', () => {
    const component = shallow(<SimpleAnimation fade />);
    component.instance().opacityAnimation = 'test';

    expect(component.instance().getOpacity()).toEqual({ opacity: 'test' });
  });

  it('should return empty array when this.getMovementTransform() is called and props.direction does not exist', () => {
    const component = shallow(<SimpleAnimation direction={null} />);
    expect(component.instance().getMovementTransform()).toEqual([]);
  });

  it('should return empty array when this.getMovementTransform() is called and props.movementType does not exist', () => {
    const component = shallow(<SimpleAnimation movementType={null} />);
    expect(component.instance().getMovementTransform()).toEqual([]);
  });

  it('should return empty array when this.getStaticTransform() is called and props.staticType does not exist', () => {
    const component = shallow(<SimpleAnimation staticType={null} />);
    expect(component.instance().getStaticTransform()).toEqual([]);
  });

  it('should return array with object with scale value when this.getStaticTransform() is called and props.staticType exists', () => {
    const component = shallow(<SimpleAnimation staticType="bounce" />);
    component.instance().staticAnimation = 'test';

    expect(component.instance().getStaticTransform()).toEqual([{ scale: 'test' }]);
  });

  it('should render null when props.children does not exist', () => {
    const component = create(<SimpleAnimation />).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('should render SimpleAnimation with children when props.children exists', () => {
    const component = create(
      <SimpleAnimation>
        <Text>test</Text>
      </SimpleAnimation>
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
});
