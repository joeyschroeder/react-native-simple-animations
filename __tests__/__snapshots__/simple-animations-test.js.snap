// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SimpleAnimation /> should render SimpleAnimation with children when props.children exists 1`] = `
<View
  style={
    Object {
      "opacity": 0,
      "transform": Array [],
    }
  }
>
  <Text>
    test
  </Text>
</View>
`;

exports[`<SimpleAnimation /> should render null when props.children does not exist 1`] = `null`;
